ramp4-pcar4/ramp4-pcar4
# Implement clean fixture removal

## Closes #947

## Changes in this PR
- [FIX] All fixtures can be removed and will clean up properly
- [FIX] Updated all `watch` usages in components to the imperative `$watch()` instead to allow for proper removal
- [FIX] Legend items will no longer be stuck in placeholder state due to race condition
    - Also fixed bug where legend config parser was directly mutating the legend config in the store 
- [FIX] Add scrollguard config to schema
- [FEAT] Added `remove` method to `PanelAPI` that will remove registered panels
- [FEAT] Added RAMP reference as `window.debugInstance` in `demos/starter-scripts/main.js` for debugging
- [FEAT] MapAPI changes
	- Split old `reloadMap` method into `destroyMapView` and `createMapView` - these protected methods will be called in sequence in `setBasemap`
	- Added `destroyMap` protected method that first calls `destroyMapView`, then destroys the ESRI map and fires the `MAP_DESTROYED` event
	- Added new `reloadMap(rampConfig, targetDiv)` public method that first calls `destroyMap`, and then calls `createMap` with the given parameters

## Comments
- Appbar and mapnav buttons are not removed when fixtures are removed
	- Removing appbar buttons is blocked by #882
	- The mapnav api should be enhanced to follow the design of the appbar after so it can add/remove mapnav buttons as fixtures are added/removed
- Fixtures will not removed globally registered components (e.g. appbar buttons) when removed
	- These components will still be available after the fixture is removed and will be reused when the fixture is added back
	- Note that some fixtures (e.g. overviewmap) will mount components onto the Vue app instead of registering them - these components are properly destroyed when the fixture is removed
- I am currently using a hack in `details/index.ts` to delay the details store removal so that the layers screen store watcher doesn't trigger as it is closing (more details in the comment)
	- Is there a better way to do this?

## [Demo](http://ramp4-app.azureedge.net/demo/users/sharvenp/947/demos/index.html)
### Steps to test

- Try removing fixtures with: `window.debugInstance.fixture.remove(<fixture_name>)`
	- The fixture should be removed from UI, vuex store, and any other references
- Try adding the same fixture back with: `window.debugInstance.fixture.add(<fixture_name>)`
	- The fixture should be added back and should function properly
- Try calling `window.debugInstance.geo.map.reloadMap(<map config>, window.debugInstance.geo.map._targetDiv)` with a different map config
	- The map should reload with the new map config (all layers will be deleted)


---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/ramp4-pcar4/ramp4-pcar4/976)
